pkg/controller/cache.go

```diff
@@ -0,0 +1,286 @@
+	syncedCh := make(chan bool, len(c.delegates)+1)
+	cas := make([]cache.Cache, 0, len(c.delegates))
+	for _, ca := range c.delegates {
+		cas = append(cas, ca)
+	}
+	cas = append(cas, c.fallback)
+	c.lock.RUnlock()
+
+	var wg sync.WaitGroup
+	ctx, cancelFn := context.WithCancel(ctx)
+
+	for _, ca := range cas {
+		wg.Add(1)
+		go func(ca cache.Cache) {
+			defer wg.Done()
+			synced := ca.WaitForCacheSync(ctx)
+			if !synced {
+				// first unsynced cache breaks the whole wait
+				cancelFn()
+			}
+			syncedCh <- synced
+		}(ca)
+	}
+
+	wg.Wait()
+	close(syncedCh)
+	cancelFn()
+
+	// any not synced?
+	for synced := range syncedCh {
+		if !synced {
+			return false
+		}
+	}
+
+	return true
+}
+
+// IndexField adds an index with the given field name on the given object type
+// by using the given function to extract the value for that field.
+func (c *GVKRoutedCache) IndexField(ctx context.Context, obj client.Object, field string, extractValue client.IndexerFunc) error {
+	gvk, err := apiutil.GVKForObject(obj, c.scheme)
+	if err != nil {
+		return fmt.Errorf("failed to get GVK for type %T: %w", obj, err)
+	}
+
+	c.lock.RLock()
+	delegate, ok := c.delegates[gvk]
+	c.lock.RUnlock()
+
+	if ok {
+		return delegate.IndexField(ctx, obj, field, extractValue)
+	}
+
+	return c.fallback.IndexField(ctx, obj, field, extractValue)
+}
+
+// cachedRoutedClient wraps a client and routes read requests by GVK to a cache.
+type cachedRoutedClient struct {
+	client.Client
+
+	scheme *runtime.Scheme
+	cache  *GVKRoutedCache
+}
+
+func (c *cachedRoutedClient) Get(ctx context.Context, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error {
+	gvk, err := apiutil.GVKForObject(obj, c.scheme)
+	if err != nil {
+		return fmt.Errorf("failed to get GVK for type %T: %w", obj, err)
+	}
+
+	c.cache.lock.RLock()
+	delegate, ok := c.cache.delegates[gvk]
+	c.cache.lock.RUnlock()
+
+	if ok {
+		return delegate.Get(ctx, key, obj, opts...)
+	}
+
+	return c.Client.Get(ctx, key, obj, opts...)
+}
+
+func (c *cachedRoutedClient) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	gvk, err := apiutil.GVKForObject(list, c.scheme)
+	if err != nil {
+		return fmt.Errorf("failed to get GVK for type %T: %w", list, err)
+	}
+
+	if !strings.HasSuffix(gvk.Kind, "List") {
+		// following controller-runtime here which does not support non
+		// <Kind>List types.
+		return fmt.Errorf("non-list type %T (kind %q) passed as output", list, gvk)
+	}
+	gvk.Kind = strings.TrimSuffix(gvk.Kind, "List")
+
+	c.cache.lock.RLock()
+	delegate, ok := c.cache.delegates[gvk]
+	c.cache.lock.RUnlock()
+
+	if ok {
+		return delegate.List(ctx, list, opts...)
+	}
+
+	return c.Client.List(ctx, list, opts...)
+}
+
+// WithGVKRoutedCache returns a manager backed by a GVKRoutedCache. The client
+// returned by the manager will route read requests to cached GVKs.
+func WithGVKRoutedCache(c *GVKRoutedCache, mgr controllerruntime.Manager) controllerruntime.Manager {
```

Comment:
Do we need to return a whole manager here? Could we return a smaller interface (e.g. Cluster)?

Reply:
yes, we do. We pass the manager to controller constructors everywhere in Crossplane. So we need this manager wrapper. It would need bigger changes to reduce it to `Cluster`.